### PR TITLE
Fix swift package init --help printout

### PR DESCRIPTION
**Pull request by miggs597**
Depends on this [PR](https://github.com/apple/swift-argument-parser/pull/300)

Fixes `swift package init --help` to only show help for `init`
This is a major UI improvement as it allows the user to focus on the help that is solely relevant to `init`, and not get overwhelmed with other options.

rdar://76116106

### CODE REVIEW

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@natecook1000 ptal

**Review by abertelrud — APPROVED** [review]
Thanks @miggs597, this is great!

So with the custom mirror, does this still depend on the ArgumentParser PR?  Or are both needed, i.e. even with the ArgumentParser PR, will the custom mirror continue to be needed?

File: Sources/Commands/SwiftPackageTool.swift [hunk 1]
```diff
@@ -190,14 +190,20 @@ extension SwiftPackageTool {
...
-    struct Init: SwiftCommand {
+    struct Init: SwiftCommand, CustomReflectable {
+        var customMirror: Mirror {
+            return Mirror(self, children: ["initMode": self._initMode, "packageName": self._packageName as Any])
+        }
+        
         static let configuration = CommandConfiguration(
             abstract: "Initialize a new package")
+        
+        static let includeSuperCommandInHelp = false
```
Comment (natecook1000) [inline Sources/Commands/SwiftPackageTool.swift:201]: We decided against requiring this bit for the ArgumentParser change.
```suggestion
```

**Comment by miggs597** [comment]
To help us catch any possible regressions it would be nice to leverage `AssertHelp`. 
[PR](https://github.com/apple/swift-argument-parser/pull/303)

File: Sources/Commands/SwiftPackageTool.swift [hunk 2]
```diff
@@ -908,7 +914,7 @@ extension SwiftPackageTool {
         @OptionGroup()
         var swiftOptions: SwiftToolOptions
 
-        @Argument()
+        @Argument(help: "generate-bash-script | generate-zsh-script |\ngenerate-fish-script | list-dependencies | list-executables")
```
Comment (tomerd) [inline Sources/Commands/SwiftPackageTool.swift:917]: Can the argument parser library enumerate the enum options instead of having to explicitly spell them like this? cc @natecook1000?

File: Sources/Commands/SwiftPackageTool.swift [hunk 3]
```diff
@@ -190,22 +190,25 @@ extension SwiftPackageTool {
...
         var packageName: String?
-
+        
+        public init() {}
+        
         func run(_ swiftTool: SwiftTool) throws {
-            // FIXME: Error handling.
-            let cwd = localFileSystem.currentWorkingDirectory!
+            guard let cwd = localFileSystem.currentWorkingDirectory else {
+                throw StringError("Could not find the current working directory")
```
Comment (tomerd) [inline Sources/Commands/SwiftPackageTool.swift:210]: throw InternalError in such case, as it would assert in debug mode which is what we want in this failure case

File: Sources/Commands/SwiftPackageTool.swift [hunk 4]
```diff
@@ -190,22 +190,25 @@ extension SwiftPackageTool {
         }
     }
 
-    struct Init: SwiftCommand {
-        static let configuration = CommandConfiguration(
+    public struct Init: SwiftCommand {
```
Comment (tomerd) [inline Sources/Commands/SwiftPackageTool.swift:193]: why does this need to be public?

File: Tests/CommandsTests/PackageToolTests.swift [hunk 5]
```diff
@@ -48,6 +49,22 @@ final class PackageToolTests: XCTestCase {
         XCTAssert(stdout.contains("Swift Package Manager"), "got stdout:\n" + stdout)
     }
     
+    func testInitHelp() throws {
+        AssertHelp(for: Commands.SwiftPackageTool.Init.self, equals: """
+            OVERVIEW: Initialize a new package
+
+            USAGE: init [<options>] --enable-index-store
```
Comment (tomerd) [inline Tests/CommandsTests/PackageToolTests.swift:56]: --enable-index-store ?

File: Sources/Commands/SwiftPackageTool.swift (hunk 4 above)
Comment (miggs597) [reply Sources/Commands/SwiftPackageTool.swift:193]: If we want to use `AssertHelp` we need access to the commands.

Comment (tomerd) [reply Sources/Commands/SwiftPackageTool.swift:193]: you can use `@testable import` to use test internal structures and functions. that is better than exposing more surface area in the public API.

File: Tests/CommandsTests/PackageToolTests.swift (hunk 5 above)
Comment (tomerd) [reply Tests/CommandsTests/PackageToolTests.swift:56]: ` but it does go away when they're all transitioned.` what does that mean?

Comment (tomerd) [reply Tests/CommandsTests/PackageToolTests.swift:56]: another point is that I dont think its scalable to string match the the entire output of the `--help` for each command. did you intend to add such tests to all commands or just this one as an indicator fo the rest?

Comment (miggs597) [reply Tests/CommandsTests/PackageToolTests.swift:56]: It doesn't have to be for all the commands, I think it would be enough to have tests for the most popular commands: e.g. `swift run`, `swift build`, `swift test` 

`--enable-index-store` goes away once all the commands have the option group hidden. In my initial testing that's when I noticed that it would go away.

**Comment by miggs597** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

File: Package.swift [hunk 6]
```diff
@@ -358,7 +358,7 @@ if ProcessInfo.processInfo.environment["SWIFTCI_USE_LOCAL_DEPS"] == nil {
         // The 'swift-argument-parser' version declared here must match that
         // used by 'swift-driver' and 'sourcekit-lsp'. Please coordinate
         // dependency version changes here with those projects.
-        .package(url: "https://github.com/apple/swift-argument-parser.git", .upToNextMinor(from: "0.3.1")),
+        .package(url: "https://github.com/apple/swift-argument-parser.git", .upToNextMinor(from: "0.4.3")),
```
Comment (tomerd) [inline Package.swift:361]: please work @shahmishal to also update the toolchain CI definition

File: Sources/Commands/Options.swift [hunk 7]
```diff
@@ -242,13 +242,28 @@ public struct SwiftToolOptions: ParsableArguments {
...
-        guard let enable = indexStoreEnable else { return .auto }
-        return enable ? .on : .off
+        switch indexStoreMode {
+        case .autoIndexStore:
+            return .auto
+        case .enableIndexStore:
+            return .on
+        case .disableIndexStore:
+            return .off
+        }
```
Comment (tomerd) [inline Sources/Commands/Options.swift:266]: this changes is a good one but we should do it in a separate PR

File: Sources/Commands/Options.swift [hunk 8]
```diff
@@ -242,13 +242,28 @@ public struct SwiftToolOptions: ParsableArguments {
...
+    // if true, false aren't enough and a third state is needed
+    // nil should not be the goto. Instead create an enum
+    enum StoreMode: String, EnumerableFlag {
+        case autoIndexStore
+        case enableIndexStore
+        case disableIndexStore
+    }
+    
+    @Flag(help: "Enable or disable indexing-while-building feature")
+    var indexStoreMode: StoreMode = .autoIndexStore
```
Comment (tomerd) [inline Sources/Commands/Options.swift:255]: this changes is a good one but we should do it in a separate PR

File: Tests/CommandsTests/PackageToolTests.swift [hunk 9]
```diff
@@ -12,7 +12,7 @@ import XCTest
 import Foundation
 
 import TSCBasic
-import Commands
+@testable import Commands
```
Comment (tomerd) [inline Tests/CommandsTests/PackageToolTests.swift:15]: used?

File: Package.swift (hunk 6 above)
Comment (shahmishal) [reply Package.swift:361]: https://github.com/apple/swift/pull/37113

**Comment by miggs597** [comment]
[Updated Toolchain](https://github.com/apple/swift/pull/37113)
@swift-ci please smoke test

**Comment by miggs597** [comment]
apple/swift/pull/37113
@swift-ci please smoke test

File: Package.swift (hunk 6 above)
Comment (tomerd) [reply Package.swift:361]: looks like this also needs to be coordinated with https://github.com/apple/swift-driver/blob/main/Package.swift#L134

cc @artemcm

**Comment by miggs597** [comment]
apple/swift-driver#534
apple/sourcekit-lsp#374
apple/swift#37113
@swift-ci please smoke test

**Comment by miggs597** [comment]
apple/swift-driver#534
apple/sourcekit-lsp#374
@swift-ci please smoke test

**Comment by miggs597** [comment]
apple/swift-driver#534
apple/sourcekit-lsp#374
@swift-ci please smoke test

File: Sources/Commands/Options.swift (hunk 8 above)
Comment (miggs597) [reply Sources/Commands/Options.swift:255]: This is what fixes `USAGE: init [<options>] --enable-index-store` preventing the index store from being printed out, should it still be another PR?

Comment (tomerd) [reply Sources/Commands/Options.swift:255]: yes please

**Comment by miggs597** [comment]
apple/swift-driver#534
apple/sourcekit-lsp#374
@swift-ci please smoke test
